polygeist/test/polygeist-opt/memoryAccess.mlir

```diff
@@ -7,43 +7,114 @@
-//      The underlying value of the accessor subscript is not immediately evident.
+// COM: Test 1-dim accessor memory access.
+//      The underlying values of the accessor subscript operations are affine expression.
 // CHECK-LABEL: test_tag: test1a_load1
 // CHECK: matrix:
 // CHECK-NEXT: 2
+// CHECK-LABEL: test_tag: test1a_load2
+// CHECK: matrix:
+// CHECK-NEXT: 3
+// CHECK-LABEL: test_tag: test1a_load3
+// CHECK: matrix:
+// CHECK-NEXT: 4
+// CHECK-LABEL: test_tag: test1a_load4
+// CHECK: matrix:
+// CHECK-NEXT: 1
+// CHECK-LABEL: test_tag: test1a_load5
+// CHECK: matrix:
+// CHECK-NEXT: 2
+// CHECK-LABEL: test_tag: test1a_load6
+// CHECK: matrix:
+// CHECK-NEXT: 1
+// CHECK-LABEL: test_tag: test1a_load7
+// CHECK: matrix:
+// CHECK-NEXT: -1
 func.func @test1a(%acc : memref<?x!sycl_accessor_1_f32_rw_gb, 4>) {
   %alloca = memref.alloca() : memref<1x!sycl_id_1>
-  %alloca_0 = memref.alloca() : memref<1x!sycl_id_1>
   %cast = memref.cast %alloca : memref<1x!sycl_id_1> to memref<?x!sycl_id_1>
-  %cast_0 = memref.cast %alloca : memref<1x!sycl_id_1> to memref<?x!sycl_id_1>
   %id = memref.memory_space_cast %cast : memref<?x!sycl_id_1> to  memref<?x!sycl_id_1, 4>
-  %c1 = arith.constant 2 : index
-
+  %c1_i32 = arith.constant 1 : i32
+  %c2_i32 = arith.constant 2 : i32
+  %c3_i32 = arith.constant 3 : i32
+  %c4_i32 = arith.constant 4 : i32
+  %c1_i64 = arith.extsi %c1_i32 : i32 to i64
+  %c2_i64 = arith.extsi %c2_i32 : i32 to i64
+  %c3_i64 = arith.extsi %c3_i32 : i32 to i64
+  %c4_i64 = arith.extsi %c4_i32 : i32 to i64
+  %negc1_i64 = arith.constant -1 : i64
+  
   affine.for %ii = 0 to 64 {
-    %i = arith.muli %ii, %c1 : index
-    sycl.constructor @id(%id, %i) {MangledFunctionName = @dummy} : (memref<?x!sycl_id_1, 4>, index)
-    %val = affine.load %alloca[0] : memref<1x!sycl_id_1>
-    affine.store %val, %alloca_0[0] : memref<1x!sycl_id_1>
-    %subscr = sycl.accessor.subscript %acc[%cast_0] : (memref<?x!sycl_accessor_1_f32_rw_gb, 4>, memref<?x!sycl_id_1>) -> memref<?xf32, 4>
-    %load1 = affine.load %subscr[0] {tag = "test1a_load1"} : memref<?xf32, 4>
+    %index_cast = arith.index_cast %ii : index to i64
+
+    // i*2
+    %mul1 = arith.muli %index_cast, %c2_i64 : i64
+    sycl.constructor @id1(%id, %mul1) {MangledFunctionName = @dummy} : (memref<?x!sycl_id_1, 4>, i64)
+    %subscr1 = sycl.accessor.subscript %acc[%cast] : (memref<?x!sycl_accessor_1_f32_rw_gb, 4>, memref<?x!sycl_id_1>) -> memref<?xf32, 4>
+    %load1 = affine.load %subscr1[0] {tag = "test1a_load1"} : memref<?xf32, 4>
+
+    // (i*3)+1
+    %mul2 = arith.muli %index_cast, %c3_i64 : i64
+    %add2 = arith.addi %mul2, %c1_i64 : i64
+    sycl.constructor @id1(%id, %add2) {MangledFunctionName = @dummy} : (memref<?x!sycl_id_1, 4>, i64)
+    %subscr2 = sycl.accessor.subscript %acc[%cast] : (memref<?x!sycl_accessor_1_f32_rw_gb, 4>, memref<?x!sycl_id_1>) -> memref<?xf32, 4>
+    %load2 = affine.load %subscr2[0] {tag = "test1a_load2"} : memref<?xf32, 4>
+
+    // (i*4)*1
```

Comment:
is it going to work if `(i*4)*2`, where both multiplier are not 1.

Reply:
No only works when a multiplier is one (it returns the other one). As mentioned in the previous response we need a better solution, will fix it.

Reply:
Technically, we should not run into this case if canonicalization is run.